src/main.rs

```diff
@@ -13,20 +18,33 @@ fn main() {
 
     // Parse command line arguments
     let args: Vec<String> = std::env::args().collect();
-    if args.len() < 3 {
-        eprintln!("Usage: {} <input wave filename> <output text filename>", args[0]);
+    if args.len() < 5 {
+        eprintln!("Usage: {} <input wave filename> <output text filename> <modulation freq (Hz)> <width (s)>", args[0]);
         return
     }
 
     // Open the input wave file
     let mut reader = hound::WavReader::open(&args[1]).unwrap();
     let spec = reader.spec();
-    let channels = spec.channels;
+    let channels = spec.channels as usize;
+    let output_file = &args[2];
+    let mod_freq = args[3].parse::<f32>().expect("Modulation Frequency input is wrong! Must be a floating point");
+    let width = args[4].parse::<f32>().expect("Width input is wrong! Must be a floating point");
+    if spec.bits_per_sample!=16 {
+        eprintln!("Bit depth must be 16 bit! Bit depth of the current song: {}", spec.bits_per_sample);
```

Comment:
I see that you have some error handling for when a wav file is not 16 bit depth. 
With high quality audio files, 24 bit depth is very common. While using test audio files I had on hand this was an unforeseen issue I ran into. Here some modifications of your code that could help your program accept both 16 and 24 bit depth. 


let mut reader = hound::WavReader::open(&args[1]).unwrap();
let spec = reader.spec();
let channels = spec.channels as usize;

if spec.bits_per_sample != 16 && spec.bits_per_sample != 24 {
    eprintln!("Bit depth must be 16 bit or 24 bit! Bit depth of the current song: {}", spec.bits_per_sample);
    return
}

let block_size: usize = 32;
let mut vibrato_filter = Vibrato::new(spec.sample_rate as f32, mod_freq, width, channels);
let mut writer: hound::WavWriter<BufWriter<File>> = hound::WavWriter::create(output_file, spec).expect("Unable to create file");

if spec.bits_per_sample == 16 {
    while let Ok(block) = reader.samples::<i16>().take(block_size*channels).collect::<Result<Vec<_>, _>>() {
        process_block(&block, &channels, &mut vibrato_filter, &mut writer);
        if block.len() < block_size*channels as usize { break }
    }
} else if spec.bits_per_sample == 24 {
    while let Ok(block) = reader.samples::<i32>().take(block_size*channels).collect::<Result<Vec<_>, _>>() {
        let block: Vec<i32> = block.into_iter().map(|sample| sample >> 8).collect(); // Shift to get 24-bit sample
        process_block(&block, &channels, &mut vibrato_filter, &mut writer);
        if block.len() < block_size*channels as usize { break }
    }
}

fn process_block<T: Sample + hound::Sample>(block: &Vec<T>, channels: &usize, vibrato_filter: &mut Vibrato, writer: &mut hound::WavWriter<BufWriter<File>>) {
    let mut process_block = ProcessBlocks::new(block, channels);
    let (input_address, mut output_address) = process_block.create_and_write_addresses();
    vibrato_filter.process(&input_address, &mut output_address);
    process_block.write_output_samples(writer).unwrap();
}

Reply:
Thanks a lot for the in-depth answer TeAiris! Actually that's a very good way to downsample 24 to 16 to do the operations written for 16. I'll implement this in the code